beta/nncf/tensorflow/initialization.py

```diff
@@ -0,0 +1,157 @@
+"""
+ Copyright (c) 2021 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from functools import partial
+from contextlib import contextmanager
+from typing import Dict
+from typing import Union
+
+import tensorflow as tf
+
+from nncf.common.utils.progress_bar import ProgressBar
+from nncf.common.initialization import NNCFDataLoader
+from nncf.config.structure import BNAdaptationInitArgs
+from nncf.config.config import NNCFConfig
+from beta.nncf.tensorflow.graph.metatypes.keras_layers import TFBatchNormalizationLayerMetatype
+from beta.nncf.tensorflow.graph.metatypes.matcher import get_keras_layer_metatype
+
+
+class InitializingDataLoader(NNCFDataLoader):
+    """
+    This class wraps the tf.data.Dataset class.
+    This is required for proper initialization of certain compression algorithms.
+    """
+
+    def __init__(self, data_loader: tf.data.Dataset):
+        self._data_loader = data_loader
+
+    @property
+    def batch_size(self) -> int:
+        for (x, _) in self._data_loader:
+            batch_size = x.shape[0]
+            return batch_size
+
+    def __iter__(self):
+        return iter(self._data_loader)
+
+
+class DataLoaderBNAdaptationRunner:

```

Comment:
Can we simplify this class? What is the reason for context managers? Why do we need so many internal methods?